vllm/spec_decode/draft_model_runner.py

```diff
@@ -115,8 +120,228 @@ def update_model_input(
+        #                                 last_output)
+
+        # Get output tokens GPU tensor
+        sampled_token_ids = last_output.sampled_token_ids
+        assert sampled_token_ids is not None
+
+        # Update attn_metadata
+        attn_metadata = model_input.attn_metadata
+        assert isinstance(attn_metadata, FlashAttentionMetadata)
+        self._update_flash_attn_metadata(attn_metadata, num_seqs, num_queries)
+
+        # Update GPU tensors
+        ops.advance_step(num_seqs=num_seqs,
+                         num_queries=num_queries,
+                         block_size=self.block_size,
+                         input_tokens=model_input.input_tokens,
+                         sampled_token_ids=sampled_token_ids,
+                         input_positions=model_input.input_positions,
+                         seq_lens=attn_metadata.seq_lens_tensor,
+                         slot_mapping=attn_metadata.slot_mapping,
+                         block_tables=attn_metadata.block_tables)
+
+        # Update sampling_metadata
+        sampling_metadata = model_input.sampling_metadata
+        self._update_sampling_metadata(sampling_metadata, num_seqs,
+                                       num_queries)
+
+        # Create new input
+        new_model_input = self._model_input_cls(
+            input_tokens=model_input.input_tokens,
+            input_positions=model_input.input_positions,
+            attn_metadata=attn_metadata,
+            seq_lens=attn_metadata.seq_lens,
+            query_lens=model_input.query_lens,
+            lora_mapping=model_input.lora_mapping,
+            lora_requests=model_input.lora_requests,
+            multi_modal_kwargs=model_input.multi_modal_kwargs,
+            sampling_metadata=model_input.sampling_metadata,
+            is_prompt=False,
+        )
+
+        if log_advance_input:
+            print("NEW INPUT: ")
+            print("  input_tokens = {}".format(new_model_input.input_tokens))
+            print("  input_positions = {}".format(
+                new_model_input.input_positions))
+            print("  seq_lens = {}".format(new_model_input.seq_lens))
+            print("  query_lens = {}".format(new_model_input.query_lens))
+            print("  attn_metadata:")
+            print("    seq_lens_tensor: {}".format(
+                attn_metadata.seq_lens_tensor))
+            print("    slot_mapping: {}".format(attn_metadata.slot_mapping))
+            print("    block_tables: {}".format(attn_metadata.block_tables))
+
+        return new_model_input
+
+    def _can_use_advance_step(self):
+        if not enable_advance_step:
+            return False
+
+        # TODO: Add support for other attn backends
+        if self.attn_backend.get_name() != "flash-attn":
+            return False
+
+        # TODO: Understand what is needed for LORA support
+        if self.lora_config:
+            return False
+
+        return True
```

Comment:
We could put this checker here: https://github.com/vllm-project/vllm/blob/main/vllm/spec_decode/spec_decode_worker.py#L133-L135, so that:
1. We use this draft model runner with advance step only when applicable (e.g., TP=1, FlashAtten, etc).
2. We can simplify the implementation of draft model runner by only supporting `enable_advance_step` case.

Reply:
Oh I see, this is actually a good idea, since the non GPU advance_step is simply the usual runner :)

Reply:
Fixed based on our discussion

Reply:
I added a condition to allow the draft_model_runner to run only when the request has only decodes already. This allowed to remove all of the other code, so it only has the GPU advance_step now and it is much simpler. Thanks for the suggestion!